src/LDATaskS3.cpp

```diff
@@ -0,0 +1,385 @@
+      // load the next LDAStatistics from S3
+      while (true) {
+        std::this_thread::sleep_for(std::chrono::milliseconds(1));
+        if (upload_lock_indicators[cur_train_idx - start] == -1) {
+          upload_lock_indicators[cur_train_idx - start] = 1;
+          break;
+        }
+      }
+
+      obj_id_str =
+          std::to_string(hash_f(std::to_string(cur_train_idx).c_str())) +
+          "-LDA";
+      start_time_benchmark = get_time_ms();
+      std::ostringstream* s3_obj =
+          s3_client->s3_get_object_ptr(obj_id_str, config.get_s3_bucket());
+
+      upload_lock_indicators[cur_train_idx - start] = -1;
+      time_download += (get_time_ms() - start_time_benchmark) / 1000.0;
+
+      // create the LDAModel
+      model.reset(new LDAModel(s3_obj->str().c_str()));
+
+      delete s3_obj;
+
+      // load the pre-cached token indices from server
+      slice_indices_mem = psint->get_slices_indices(cur_train_idx - start);
+      load_serialized_indices(slice_indices_mem);
+      delete slice_indices_mem;
+
+      num_runs = 3000 / config.get_slice_size() + 1;
+      cur = 0;
+
+      continue;
+    }
+
+    cur += 1;
+
+    start_time_benchmark = get_time_ms();
+    // load one word slice from server
+    char* partial_model =
+        psint->get_lda_model(to_receive_size, uncompressed_size);
+    time_get_model += (get_time_ms() - start_time_benchmark) / 1000.0;
+
+    start_time_benchmark = get_time_ms();
+    // update current LDAModel
+    model->update_model(partial_model, to_receive_size, uncompressed_size,
+                        psint->slice_id);
+    time_create_model += (get_time_ms() - start_time_benchmark) / 1000.0;
+    num_runs = model->get_V() / config.get_slice_size();
+
+    // sampling phase
+    char* gradient_mem;
+    uint32_t gradient_size;
+    start_time_benchmark = get_time_ms();
+    gradient_mem = model->sample_model(
+        total_sampled_tokens, slice_indices[psint->slice_id], gradient_size);
+    time_sample += (get_time_ms() - start_time_benchmark) / 1000.0;
+
+    // push the update to server
+    try {
+      start_time_benchmark = get_time_ms();
+      push_gradient(gradient_mem, total_sampled_tokens, total_sampled_docs,
+                    gradient_size);
+
+      total_sampled_docs = 0;
+
+      time_update += (get_time_ms() - start_time_benchmark) / 1000.0;
+    } catch (...) {
+      std::cout << "[WORKER] "
+                << "Worker task error doing put of gradient"
+                << "\n";
+      exit(-1);
+    }
+
+    count++;
+
+    if (count % 10 == 0 && !printed_rate) {
+      auto elapsed_ms = get_time_ms() - start_time;
+      float elapsed_sec = elapsed_ms / 1000.0;
+      if (elapsed_sec > (2 * 60)) {
+        printed_rate = true;
+        std::cout << "Update rate/sec last 2 mins: "
+                  << (1.0 * count / elapsed_sec) << std::endl;
+      }
+    }
+    int since_start_sec = (get_time_ms() - start_time) / 1000;
+    if (since_start_sec > benchmark_time) {
```

Comment:
Same thing, this should go into separate function

Reply:
^^Code refactoring has been done. Please let me know if it remains messy.